File: packages/kotlin/src/test/kotlin/com/luxo/client/ErrorTest.kt
package com.luxo.client

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

class LuxoErrorTest {

    @Test
    fun `constructor sets all properties`() {
        val data = buildJsonObject { put("field", "email") }
        val err = LuxoError(
            error = "NotFound",
            code = 404,
            message = "user not found",
            traceId = "abc-123",
            data = data,
            developmentCause = "database constraint",
        )

        assertEquals("NotFound", err.error)
        assertEquals(404, err.code)
        assertEquals("user not found", err.message)
        assertEquals("abc-123", err.traceId)
        assertEquals(data, err.data)
        assertEquals("database constraint", err.developmentCause)
    }

    @Test
    fun `traceId defaults to null`() {
        val err = LuxoError(
            error = "Unauthorized",
            code = 401,
            message = "invalid token",
        )

        assertNull(err.traceId)
    }

    @Test
    fun `is an Exception`() {
        val err = LuxoError("InternalError", 500, "something went wrong")
        val exception: Exception = err
        assertTrue(exception is LuxoError)
        assertEquals("something went wrong", err.message)
    }

    @Test
    fun `toString format`() {
        val err = LuxoError("BadRequest", 400, "missing field")
        assertEquals("LuxoError(BadRequest, 400, missing field)", err.toString())
    }

    @Test
    fun `can be thrown and caught`() {
        try {
            throw LuxoError("Conflict", 409, "duplicate entry", "trace-xyz")
        } catch (e: LuxoError) {
            assertEquals("Conflict", e.error)
            assertEquals(409, e.code)
            assertEquals("duplicate entry", e.message)
            assertEquals("trace-xyz", e.traceId)
        }
    }
}

File: packages/dart/test/error_test.dart
import 'package:test/test.dart';
import 'package:luxo_client/src/error.dart';

void main() {
  group('LuxoError', () {
    test('constructor sets all fields', () {
      final err = LuxoError('NotFound', 404, 'user not found', 'trace-abc-123');
      expect(err.error, equals('NotFound'));
      expect(err.code, equals(404));
      expect(err.message, equals('user not found'));
      expect(err.traceId, equals('trace-abc-123'));
    });

    test('traceId is optional (null)', () {
      final err = LuxoError('ServerError', 500, 'internal error');
      expect(err.traceId, isNull);
    });

    test('carries structured data and development cause', () {
      final err = LuxoError(
          'BadRequest',
          400,
          'bad',
          'trace',
          {
            'param': 'email',
          },
          'validation failed');
      expect(err.data, equals({'param': 'email'}));
      expect(err.cause, equals('validation failed'));
    });

    test('toString format', () {
      final err = LuxoError('Unauthorized', 401, 'invalid token');
      expect(
        err.toString(),
        equals('LuxoError(Unauthorized, 401, invalid token)'),
      );
    });

    test('toString with empty message', () {
      final err = LuxoError('Unknown', 0, '');
      expect(err.toString(), equals('LuxoError(Unknown, 0, )'));
    });

    test('implements Exception', () {
      final err = LuxoError('Test', 0, 'msg');
      expect(err, isA<Exception>());
    });

    test('can be thrown and caught', () {
      expect(
        () => throw LuxoError('BadRequest', 400, 'missing field'),
        throwsA(isA<LuxoError>().having((e) => e.code, 'code', 400)),
      );
    });
  });
}
